mongo/integration/sessions_mongocryptd_prose_test.go

```diff
@@ -0,0 +1,169 @@
+		"mongocryptd",
+		"--port", strconv.Itoa(port),
+	}
+
+	p.cmd = exec.Command(args[0], args[1:]...) //nolint:gosec
+	p.cmd.Stderr = p.cmd.Stdout
+
+	return p.cmd.Start()
+}
+
+// close will kill the underlying process on the command.
+func (p *mongocryptdProcess) close() error {
+	if p.cmd.Process == nil {
+		return nil
+	}
+
+	if err := p.cmd.Process.Kill(); err != nil {
+		return fmt.Errorf("failed to terminate mongocryptd process: %w", err)
+	}
+
+	// Release instead of wait to avoid blocking in CI.
+	err := p.cmd.Process.Release()
+	if err != nil {
+		return fmt.Errorf("failed to release mongocryptd: %w", err)
+	}
+
+	return nil
+}
+
+func TestSessionsMongocryptdProse(t *testing.T) {
+	const mongocryptdPort = 27022
+
+	// Monitor the lsid value on commands. If an operation run in any
+	// subtests contains an lsid, then the Go Driver wire message
+	// construction has incorrectly interpreted that
+	// LogicalSessionTimeoutMinutes was returned by the server on handshake.
+	cmdMonitor := &event.CommandMonitor{
+		Started: func(_ context.Context, evt *event.CommandStartedEvent) {
+			_, err := evt.Command.LookupErr("lsid")
+			assert.ErrorIs(t, err, bsoncore.ErrElementNotFound)
+		},
+	}
+
+	uri := &url.URL{
+		Scheme: "mongodb",
+		Host:   net.JoinHostPort("localhost", strconv.Itoa(mongocryptdPort)),
+	}
+
+	mtOpts := mtest.NewOptions().
+		MinServerVersion("4.2").
+		Topologies(mtest.ReplicaSet, mtest.Sharded).
+		CreateCollection(false).
+		CreateClient(false)
+
+	// Create a new instance of mtest (MongoDB testing framework) for this
+	// test and configure it to control server versions.
+	mt := mtest.New(t, mtOpts)
+	mt.Cleanup(mt.Close)
+
+	proc := mongocryptdProcess{}
+
+	// Start a mongocryptd server.
+	err := proc.start(mongocryptdPort)
+	require.NoError(t, err, "failed to create a mongocryptd process: %v", err)
+
+	t.Cleanup(func() {
+		err := proc.close()
+		require.NoError(t, err, "failed to close mongocryptd: %v", err)
+	})
+
+	clientOpts := options.
+		Client().
+		ApplyURI(uri.String()).
+		SetMonitor(cmdMonitor)
+
+	ctx := context.Background()
+
+	client, err := mongo.Connect(ctx, clientOpts)
+	require.NoError(t, err, "could not connect to mongocryptd: %v", err)
+
+	t.Cleanup(func() {
+		err := client.Disconnect(ctx)
+		require.NoError(t, err, "mongocryptd client could not disconnect: %v", err)
+	})
+
+	mt.RunOpts("18. implicit session is ignored if connection does not support sessions", mtOpts, func(mt *mtest.T) {
+		coll := client.Database("db").Collection("coll")
+
+		// Send a read command to the server (e.g., findOne), ignoring
+		// any errors from the server response
+		mt.RunOpts("read", mtOpts, func(_ *mtest.T) {
+			_ = coll.FindOne(context.Background(), bson.D{{"x", 1}})
+		})
+
+		// Send a write command to the server (e.g., insertOne),
+		// ignoring any errors from the server response
+		mt.RunOpts("write", mtOpts, func(_ *mtest.T) {
+			_, _ = coll.InsertOne(context.Background(), bson.D{{"x", 1}})
+		})
```

Comment:
If we're trying to assert that running the operations succeeds, should we capture the error and assert that it's `nil` instead?

Reply:
InsertOne will always result in an error in this case since mongocryptd servers do not support operations. The goal of the test is to make sure that a server that does not support sessions, such as mongocryptd, does not include an lsid. All of these assertions are made in the command monitor defined on the client for this test:

```go
	cmdMonitor := &event.CommandMonitor{
		Started: func(_ context.Context, evt *event.CommandStartedEvent) {
			_, err := evt.Command.LookupErr("lsid")
			assert.ErrorIs(t, err, bsoncore.ErrElementNotFound)
		},
	}
```

Reply:
Making assertions outside of a subtest about something that happen in a subtest may lead to some very confusing test failure messaging. Additionally, it's unexpected and confusing when reading the tests.

I think we should restructure these test cases so we can make the assertions in the same subtest block where we run the operations. Consider creating a `Client` per subtest so you can set up the monitor to make assertions using the same subtest's `mtest.T`.

Reply:
The updates look good. My only additional thought is the `mt.RunOpts` subtests for "read" and "write" still seem unnecessary because they don't have any additional assertions and don't modify the `mtOpts`. The test seems like it would behave exactly the same if the calls to `FindOne` and `InsertOne` were in the "prose test 18" function block.